---
title:  GOMAP
date:   2020-08-10
layout: page
tags: GO Annotation,Bioinformatics, Iowa State University, CAFA, Singularity
---
## GOMAP

Gene Ontology Meta Annotator for Plants (GOMAP) is a high-throughput pipeline to annotate GO terms to plant protein sequences. The pipeline uses three different approaches to annotate GO terms to plant protein sequences, and compile the annotations together to generate an aggregated dataset. GOMAP uses Python code to run the component tools to generate the GO annotations, and R code to clean and aggregate the GO annotations from the component tools.

### GOMAP Resources

1. GOMAP Code:
   1. GOMAP-singularity: [github.com/Dill-PICL/GOMAP-singularity](https://github.com/Dill-PICL/GOMAP-singularity)
   2. GOMAP pipeline code: [github.com/Dill-PICL/GOMAP](https://github.com/Dill-PICL/GOMAP)
2. GOMAP Documentation: [bioinformapping.com/gomap](https://bioinformapping.com/gomap)
3. GOMAP-PlantSpecific: [github.com/bioinformapping/GOMAP-PlantSpecific](https://github.com/bioinformapping/GOMAP-PlantSpecific)
4. GOMAP Annotation Datasets: [dill-picl.org/projects/gomap/gomap-datasets](https://dill-picl.org/projects/gomap/gomap-datasets)
